src/main/java/com/google/sps/servlets/DashboardServlet.java

```diff
@@ -14,62 +14,133 @@
   
-    HashMap<String, Double> interactionPercentages = new HashMap<>();
+    HashMap<String, Double> dataToSend = new HashMap<>();
 
-    for (Field field : allFields) {
-      String property = (String) field.get(new Property()); // gets the value of the field variable
+    for (int i = 0; i < searchableProperties.length; ++i){
+      String property = searchableProperties[i];
+
+      Filter propertyFilter =  new FilterPredicate(property, FilterOperator.EQUAL, true);
+      Query filteredQuery = new Query(DBUtilities.INTERACTION_TABLE);
+      filteredQuery.setFilter(CompositeFilterOperator.and(propertyFilter, startTimestampFilter, endTimestampFilter));
 
-      Filter keyFilter =  new FilterPredicate(property, FilterOperator.EQUAL, true);
-      Query filteredQuery = new Query(DBUtilities.INTERACTION_TABLE).setFilter(keyFilter);
-    
       int numUsersInteracted = datastore.prepare(filteredQuery).countEntities();
       int totalInteractions = interactions.countEntities();
 
-      interactionPercentages.put(property, numUsersInteracted / (double) totalInteractions);
+      double percentage = totalInteractions == 0 ? 0 : numUsersInteracted / (double) totalInteractions;
+
+      dataToSend.put(property, percentage);
     }
 
-    return interactionPercentages;
+    dataToSend.put("totalInteractions", (double) interactions.countEntities());
+    
+    return dataToSend;
   }
 
+ /**
+  * Given the data to convert to json, this function returns a json
+  * representation.
+  *
+  * @param data A hashmap from a string to a double to be converted to a json format.
+  * @return A json representation of the input data.
+  */
   private String convertToJson(HashMap<String, Double> data) {
     Gson gson = new Gson();
     String json = gson.toJson(data);
     return json;
   }
+
+ /**
+  * Given the start and end timestamps requested, this function validates and sets 
+  * the instance variables.
+  *
+  * @param startTimestamp A string containing the requested start time as a UNIX timestamp
+  * @param endTimestamp A string containing the requested end time as a UNIX timestamp
+  * @return An arraylist with two entries - one for each validated start/end time
+  */
+  private ArrayList<String> validateTimestamps(String requestStartTimestamp, String requestEndTimestamp){
+
+    ArrayList<String> validatedTimestamps = new ArrayList<>();
+
+    if (requestStartTimestamp.isEmpty() || requestEndTimestamp.isEmpty()) {
+      String startTimestamp = "0";
+      String endTimestamp = String.valueOf(System.currentTimeMillis());
+
+      validatedTimestamps.add(startTimestamp);
+      validatedTimestamps.add(endTimestamp);
+    } else {  
+      String startTimestamp = requestStartTimestamp;
+      String endTimestamp = requestEndTimestamp;
+    
+      // When displaying the range of dates, the latter date is included
+      endTimestamp = String.valueOf(Long.valueOf(endTimestamp) + MILLISECONDS_IN_DAY);
```

Comment:
But when you make the query, it will actually look at the actual timestamp right, not just the day?  So if they make their request for june 3rd at 4:00, this would then include everything until june 4th at 4:00?

Reply:
no, if they choose june 3rd, the timestamp that is sent is for 12:00AM of june 3rd. That's why i add the number of milliseconds in a day to include all of that date